persistence/sql/src/main/java/org/infinispan/persistence/sql/AbstractSchemaJdbcStore.java

```diff
@@ -0,0 +1,626 @@
+import org.infinispan.marshall.protostream.impl.SerializationContextRegistry;
+import org.infinispan.persistence.jdbc.common.SqlManager;
+import org.infinispan.persistence.jdbc.common.connectionfactory.ConnectionFactory;
+import org.infinispan.persistence.jdbc.common.impl.BaseJdbcStore;
+import org.infinispan.persistence.jdbc.common.sql.BaseSqlManager;
+import org.infinispan.persistence.spi.InitializationContext;
+import org.infinispan.persistence.spi.MarshallableEntry;
+import org.infinispan.persistence.spi.MarshallableEntryFactory;
+import org.infinispan.persistence.sql.configuration.AbstractSchemaJdbcConfiguration;
+import org.infinispan.persistence.sql.configuration.SchemaJdbcConfiguration;
+import org.infinispan.protostream.ImmutableSerializationContext;
+import org.infinispan.protostream.descriptors.Descriptor;
+import org.infinispan.protostream.descriptors.EnumDescriptor;
+import org.infinispan.protostream.descriptors.FieldDescriptor;
+import org.infinispan.protostream.descriptors.GenericDescriptor;
+
+public abstract class AbstractSchemaJdbcStore<K, V, C extends AbstractSchemaJdbcConfiguration> extends BaseJdbcStore<K, V, C> {
+   protected PersistenceMarshaller persistenceMarshaller;
+   protected ImmutableSerializationContext serializationContext;
+   protected String cacheName;
+   protected Parameter[] parameters;
+   protected Parameter[] primaryParameters;
+
+   protected String keyMessageName;
+   protected String valueMessageName;
+
+   @Override
+   protected SqlManager<K, V> createSqlManager(InitializationContext ctx, C config) throws SQLException {
+      AdvancedCache<K, V> advancedCache = ctx.getCache().getAdvancedCache();
+      // We use a type as the protostream -> json conversion leaves it as a String instead of byte[]
+      MediaType jsonStringType = MediaType.fromString(MediaType.APPLICATION_JSON_TYPE + ";type=String");
+      // This seems like a bug that `withRequestMediaType` isn't injected...
+      DataConversion keyDataConversion = advancedCache.getKeyDataConversion()
+            .withRequestMediaType(jsonStringType);
+      DataConversion valueDataConversion = advancedCache.getValueDataConversion()
+            .withRequestMediaType(jsonStringType);
+
+      ComponentRegistry componentRegistry = advancedCache.getComponentRegistry();
+      componentRegistry.wireDependencies(keyDataConversion, true);
+      componentRegistry.wireDependencies(valueDataConversion, true);
+
+      parameters = generateParameterInformation(config, connectionFactory);
+      assert parameters.length != 0;
+      primaryParameters = determinePrimaryParameters(config, parameters);
+      assert primaryParameters.length != 0;
+      assert Arrays.stream(primaryParameters).allMatch(Parameter::isPrimaryIdentifier);
+
+      persistenceMarshaller = ctx.getPersistenceMarshaller();
+      // TODO: we need to figure out what to do with the context here
+//      serializationContext = persistenceMarshaller.getSerializationContext();
```

Comment:
This may or may not be needed, depending upon how we should do the serialization context stuff. I am guessing I still need to remove this and instead use the user context and if not then the system context, correct?

Reply:
> I am guessing I still need to remove this and instead use the user context and if not then the system context, correct?

Thinking about this some more, as we're only concerned with converting user types for key/values, there shouldn't be any reason that we need to consider the persistence context here. All of our internal persistence types are known. The only thing we need to validate here is that the user's Message types exist, so checking the user ctx should suffice.